backend/src/main/java/ai/verta/modeldb/experimentRun/FutureExperimentRunDAO.java

```diff
@@ -365,4 +418,362 @@ public FutureExperimentRunDAO(Executor executor, FutureJdbi jdbi, UAC uac) {
+  }
+
+  private InternalFuture<ExperimentRun> checkIfEntityAlreadyExists(
+      ExperimentRun experimentRun, Boolean isInsert) {
+    return jdbi.useHandle(
+            handle -> {
+              try {
+                String queryStr;
+                if (isInsert) {
+                  queryStr =
+                      "SELECT id FROM experiment_run ere WHERE "
+                          + " ere."
+                          + ModelDBConstants.NAME
+                          + " = :experimentRunName "
+                          + " AND ere."
+                          + ModelDBConstants.PROJECT_ID
+                          + " = :projectId "
+                          + " AND ere."
+                          + ModelDBConstants.EXPERIMENT_ID
+                          + " = :experimentId "
+                          + " AND ere."
+                          + ModelDBConstants.DELETED
+                          + " = false ";
+                } else {
+                  queryStr =
+                      "SELECT id FROM experiment_run ere WHERE "
+                          + " ere."
+                          + ModelDBConstants.ID
+                          + " = :experimentRunId "
+                          + " AND ere."
+                          + ModelDBConstants.DELETED
+                          + " = false ";
+                }
+
+                var query = handle.createQuery(queryStr);
+
+                if (isInsert) {
+                  query.bind("experimentRunName", experimentRun.getName());
+                  query.bind("projectId", experimentRun.getProjectId());
+                  query.bind("experimentId", experimentRun.getExperimentId());
+                } else {
+                  query.bind(ModelDBConstants.EXPERIMENT_RUN_ID_STR, experimentRun.getId());
+                }
+
+                long count = query.mapTo(String.class).stream().count();
+                boolean existStatus = false;
+                if (count > 0) {
+                  existStatus = true;
+                }
+
+                // Throw error if it is an insert request and ExperimentRun with same name already
+                // exists
+                if (existStatus && isInsert) {
+                  throw new AlreadyExistsException("ExperimentRun already exists in database");
+                } else if (!existStatus && !isInsert) {
+                  // Throw error if it is an update request and ExperimentRun with given name does
+                  // not exist
+                  throw new NotFoundException("ExperimentRun does not exist in database");
+                }
+              } catch (Exception ex) {
+                if (ModelDBUtils.needToRetry(ex)) {
+                  checkIfEntityAlreadyExists(experimentRun, isInsert);
+                } else {
+                  throw ex;
+                }
+              }
+            })
+        .thenCompose(unused -> InternalFuture.completedInternalFuture(experimentRun), executor);
```

Comment:
The caller is already sending the ER to you. There's no need to send it back to them. This method should return an optional.

Reply:
Updated code as per suggestions